libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2918,21 +2925,9 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 					_memberAccess.location(),
 					"\"runtimeCode\" is not available for contracts containing immutable variables."
 				);
-			if (m_currentContract)
-			{
-				// TODO in the same way as with ``new``,
-				// this is not properly detecting creation-cycles if they go through
-				// internal library functions or free functions. It will be caught at
-				// code generation time, but it would of course be better to catch it here.
 
-				m_currentContract->annotation().contractDependencies.insert(&accessedContractType.contractDefinition());
-				if (contractDependenciesAreCyclic(*m_currentContract))
-					m_errorReporter.typeError(
-						4224_error,
-						_memberAccess.location(),
-						"Circular reference for contract code access."
-					);
-			}
+			if (m_currentContract || m_currentFreeFunction)
```

Comment:
This has to be moved inside the call graph analysis.

Reply:
Accessing another contract's code is the same as using `new OtherContract`.

Reply:
so adding this to the `_memberAccess` visitor and removing the part here should be enough?

```
	TypePointer exprType = _memberAccess.expression().annotation().type;
	ASTString const& memberName = _memberAccess.memberName();

	if (auto magicType = dynamic_cast<MagicType const*>(exprType))
		if (magicType->kind() == MagicType::Kind::MetaType && (
			memberName == "creationCode" || memberName == "runtimeCode"
		))
		{
			ContractType const& accessedContractType = dynamic_cast<ContractType const&>(*magicType->typeArgument());
			m_graph.createdContracts.emplace(&accessedContractType, &_memberAccess);
		}
```

Reply:
Well, I'd personally just use a variant. It's not really that much more intrusive than `ASTNode *` because with it you just trade `get<T>` for `dynamic_cast<T>`. But that's just my preference. `ASTNode *` is still a valid choice so if you think it's better suited here I'm fine that that too. I'll just probably be suggesting extra assertions because of that :)